src/overlay/Tracker.cpp

```diff
@@ -89,61 +90,92 @@ Tracker::tryNextPeer()
         {
-            if (peersWithEnvelope.find(p) != peersWithEnvelope.end())
+            auto& p = mp.second;
+            if (p->isAuthenticated() &&
+                mPeersAsked.find(p) == mPeersAsked.end())
             {
-                mPeersToAsk.emplace_back(p);
+                constexpr int64 GROUPSIZE_MS = 500;
+                int64 plat = p->getPing().count() / GROUPSIZE_MS;
+                if (plat < curBest)
+                {
+                    candidates.clear();
+                    curBest = plat;
+                    candidates.emplace_back(p);
+                }
+                else if (curBest == plat)
+                {
+                    candidates.emplace_back(p);
+                }
             }
-            else
+        }
+    };
+
+    // build the set of peers we didn't ask yet that have this envelope
+    std::map<NodeID, Peer::pointer> newPeersWithEnvelope;
+    for (auto const& e : mWaitingEnvelopes)
+    {
+        auto const& s = mApp.getOverlayManager().getPeersKnows(e.first);
+        for (auto pit = s.begin(); pit != s.end(); ++pit)
+        {
+            auto& p = *pit;
+            if (p->isAuthenticated() &&
+                mPeersAsked.find(p) == mPeersAsked.end())
             {
-                mPeersToAsk.emplace_front(p);
+                newPeersWithEnvelope.emplace(p->getPeerID(), *pit);
             }
         }
+    }
 
-        mNumListRebuild++;
-
-        CLOG(TRACE, "Overlay")
-            << "tryNextPeer " << hexAbbrev(mItemHash) << " attempt "
-            << mNumListRebuild << " reset to #" << mPeersToAsk.size();
+    if (!newPeersWithEnvelope.empty())
+    {
+        procPeers(newPeersWithEnvelope);
+    }
+    else
+    {
+        auto& inPeers = mApp.getOverlayManager().getInboundAuthenticatedPeers();
+        auto& outPeers =
+            mApp.getOverlayManager().getOutboundAuthenticatedPeers();
+        procPeers(inPeers);
+        procPeers(outPeers);
     }
 
-    while (!peer && !mPeersToAsk.empty())
+    // pick a random element from the candidate list
+    if (!candidates.empty())
     {
-        peer = mPeersToAsk.back();
-        if (!peer->isAuthenticated())
-        {
-            peer.reset();
-        }
-        mPeersToAsk.pop_back();
+        auto dist =
+            std::uniform_int_distribution<size_t>(0, candidates.size() - 1);
+        size_t offset = dist(gRandomEngine);
+        mLastAskedPeer = candidates.at(offset);
     }
 
     std::chrono::milliseconds nextTry;
-    if (!peer)
-    { // we have asked all our peers
-        // clear mLastAskedPeer so that we rebuild a new list
-        mLastAskedPeer.reset();
+    if (!mLastAskedPeer)
+    {
+        // we have asked all our peers, reset the list and try again after a
+        // pause
+        mNumListRebuild++;
+        mPeersAsked.clear();
+
+        CLOG(TRACE, "Overlay") << "tryNextPeer " << hexAbbrev(mItemHash)
+                               << " restarting fetch #" << mNumListRebuild;
+
```

Comment:
I feel like the following if/else block should just be one statement that reads `nextTry = MS_TO_WAIT_FOR_FETCH_REPLY * std::min(MAX_REBUILD_FETCH_LIST, mNumListRebuild)`

Reply:
not from this PR, but sure I made the change

Reply:
Thanks :)